lib/cn/csched_sp3_work.c

```diff
@@ -324,46 +324,82 @@ sp3_work_wtype_split(
     struct cn_tree *tree = tn->tn_tree;
     struct kvset_list_entry *le;
     struct list_head *head;
-    uint64_t now;
+    bool splittable;
     uint kvsets;
 
     head = &tn->tn_kvset_list;
     *mark = list_last_entry_or_null(head, typeof(*le), le_link);
-    kvsets = cn_ns_kvsets(&tn->tn_ns);
-
     *action = CN_ACTION_SPLIT;
     *rule = CN_RULE_SPLIT;
 
+    kvsets = cn_ns_kvsets(&tn->tn_ns);
+
+    /* Recheck to see if this node should be split, as while awaiting
+     * an opportunity to split an rspill could have shrunk this node
+     * below the split size.
+     */
+    splittable = (cn_ns_clen(&tn->tn_ns) >= tn->tn_split_size) ||
+        (cn_ns_keys_uniq(&tn->tn_ns) >= thresh->lcomp_split_keys);
+    splittable &= (kvsets > 0);
+
     mutex_lock(&tn->tn_ss_lock);
-    now = get_time_ns();
+    if (splittable) {
+        uint spilling = atomic_read(&tn->tn_ss_spilling);
+
+        if (!tn->tn_ss_splitting) {
+            if (atomic_read(&tree->ct_split_cnt) >= thresh->split_cnt_max || jclock_ns < tree->ct_split_dly) {
+                tn->tn_ss_visits = 0;
+                kvsets = 0;
+                ev_debug(1);
+            } else if (cn_ns_kvsets(&tree->ct_root->tn_ns) >= thresh->rspill_runlen_min * 3) {
+                tn->tn_ss_visits = 0;
+                kvsets = 0;
+                ev_debug(1);
+            } else if (spilling && tn->tn_ss_visits < thresh->split_cnt_max) {
+                tn->tn_ss_visits++;
+                kvsets = 0;
+                ev_debug(1);
+            } else {
+                if (atomic_inc_return(&tree->ct_split_cnt) >= thresh->split_cnt_max)
+                    tree->ct_split_dly = jclock_ns + NSEC_PER_SEC * 3;
+
+                /* By setting tn_ss_splitting to true we are committing
+                 * to split this node despite the fact that we may not
+                 * be able to start the split right now if the node is
+                 * currently being spilled.
+                 */
+                tn->tn_ss_splitting = true;
+                tn->tn_ss_visits = 0;
+                ev_debug(1);
+            }
+        }
 
-    if (!tn->tn_ss_splitting) {
-        if (tree->ct_split_cnt >= thresh->split_cnt_max || now < tree->ct_split_dly) {
-            tn->tn_ss_visits = 0;
-            kvsets = 0;
-            ev_debug(1);
-        } else if (cn_ns_kvsets(&tree->ct_root->tn_ns) >= thresh->rspill_runlen_min * 3) {
-            tn->tn_ss_visits = 0;
-            kvsets = 0;
-            ev_debug(1);
-        } else if (tn->tn_ss_spilling && tn->tn_ss_visits < thresh->split_cnt_max) {
-            tn->tn_ss_visits++;
+        /* Defer starting the split if there are active spills to this node.
+         * This most often occurs on trees with one leaf, and rarely on trees
+         * with many leaves (but largely depends on the spill pattern).
+         */
+        if (spilling)
             kvsets = 0;
+    } else {
+        if (tn->tn_ss_splitting) {
```

Comment:
I like this better than your test patch where the split job was forced to run in sp3_dirty_node_locked() even though it didn't meet the split criteria!

Reply:
Heh, yeah, I didn't like that either...  It would have been simpler just to let cn_split() reject it, but then I really wanted to confine the policy decisions to csched/sp3...